### PR TITLE
Add optimistic option. #4542

### PR DESCRIPTION
**Pull request by Koenkk**

### CODE REVIEW

File: lib/extension/publish.js [hunk 1]
```diff
@@ -238,7 +239,7 @@ class EntityPublish extends Extension {
                         this.publishEntityState(resolvedEntity.settings.ID, msg);
                     }
 
-                    if (result && result.membersState) {
+                    if (result && result.membersState && optimistic) {
```
Comment (sjorge) [inline lib/extension/publish.js:242]: Is this also true for say a group of blinds? I can't figure it out by reading more of the code around here if it is. I think so?

Comment (Koenkk) [reply lib/extension/publish.js:242]: It depends if you setup reporting

Comment (sjorge) [reply lib/extension/publish.js:242]: Nice, I'll be setting this on my groups then.
That won't fix #3461 but it will remove one of the updates.

Comment (sjorge) [reply lib/extension/publish.js:242]: It is indeed working for groups, but the effect is weird:

with 'optimistic' set to false:
- send /set to group
- reporting updates bulb1 and bulb2
- no update on the group at all

without 'optimistic' or set to true:
- send /set to group
- update of group
- reportining updates bubl1
- update of group
- reporting updates bulb2
- update of group

Is this the expected behavior?
I would have expected the following with optimistic set to false:
- send /set to group
- reporting for bulb1
- update of group
- reporting of bubl2
- update of group

Comment (Koenkk) [reply lib/extension/publish.js:242]: Not expected behaviour, I expect what you mentioned last, can you provide the debug logging of this?

**Comment by sjorge** [comment]
My group config:

```
  '10105':
    friendly_name: livingroom/floor_light
    optimistic: false
    devices:
      - 0x00158d0003618c46/1
      - 0x00158d000361b295/1
```

This is when controlling the group via a remote:
```
Zigbee2MQTT:debug 2020-10-07 19:07:29: Received Zigbee message from 'livingroom/floor_light/remote', type 'commandOff', cluster 'genOnOff', data '{}' from endpoint 1 with groupID 10105
Zigbee2MQTT:debug 2020-10-07 19:07:29: Received Zigbee message from 'livingroom/floor_light/bulb2', type 'attributeReport', cluster 'genOnOff', data '{"onOff":0}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2020-10-07 19:07:29: Received Zigbee message from 'livingroom/floor_light/bulb1', type 'attributeReport', cluster 'genOnOff', data '{"onOff":0}' from endpoint 1 with groupID 0
Zigbee2MQTT:info  2020-10-07 19:07:31: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/remote', payload '{"action":"off","battery":60,"linkquality":60,"update":{"state":"idle"}}'
Zigbee2MQTT:info  2020-10-07 19:07:31: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb2', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":52,"state":"OFF"}'
Zigbee2MQTT:info  2020-10-07 19:07:31: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb1', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":63,"state":"OFF"}'
Zigbee2MQTT:debug 2020-10-07 19:07:37: Received Zigbee message from 'livingroom/floor_light/bulb2', type 'attributeReport', cluster 'genLevelCtrl', data '{"currentLevel":254}' from endpoint 1 with groupID 0
Zigbee2MQTT:info  2020-10-07 19:07:39: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb2', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":55,"state":"OFF"}'
Zigbee2MQTT:debug 2020-10-07 19:07:42: Received Zigbee message from 'livingroom/floor_light/bulb2', type 'attributeReport', cluster 'lightingColorCtrl', data '{"currentX":45783,"currentY":19625}' from endpoint 1 with groupID 0
Zigbee2MQTT:info  2020-10-07 19:07:44: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb2', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":55,"state":"OFF"}'
Zigbee2MQTT:debug 2020-10-07 19:07:53: Received Zigbee message from 'livingroom/floor_light/bulb2', type 'attributeReport', cluster 'lightingColorCtrl', data '{"colorTemperature":555}' from endpoint 1 with groupID 0
Zigbee2MQTT:info  2020-10-07 19:07:55: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb2', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":55,"state":"OFF"}'
```

This is when controlling via mqtt

```
Zigbee2MQTT:debug 2020-10-07 19:08:41: Received MQTT message on 'zigbee2mqtt/livingroom/floor_light/set/state' with data 'toggle'
Zigbee2MQTT:debug 2020-10-07 19:08:41: Publishing 'set' 'state' to 'livingroom/floor_light'
Zigbee2MQTT:debug 2020-10-07 19:08:41: Received Zigbee message from 'livingroom/floor_light/bulb1', type 'attributeReport', cluster 'genOnOff', data '{"onOff":1}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2020-10-07 19:08:41: Received Zigbee message from 'livingroom/floor_light/bulb2', type 'attributeReport', cluster 'genOnOff', data '{"onOff":1}' from endpoint 1 with groupID 0
Zigbee2MQTT:info  2020-10-07 19:08:43: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb1', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":55,"state":"ON"}'
Zigbee2MQTT:info  2020-10-07 19:08:43: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb2', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":57,"state":"ON"}'
```

No update of the group either way.

**Comment by Koenkk** [comment]
This seemed to conflict with the `optimistic` option which was already there for groups: https://www.zigbee2mqtt.io/information/groups.html#state-changes. This is fixed now, so the old behaviour of the optimistic for groups is in place now.

**Comment by sjorge** [comment]
Something still seems of after that change

```
Zigbee2MQTT:debug 2020-10-08 18:31:46: Received Zigbee message from 'livingroom/floor_light/remote', type 'commandOn', cluster 'genOnOff', data '{}' from endpoint 1 with groupID 10105
Zigbee2MQTT:debug 2020-10-08 18:31:46: Received Zigbee message from 'livingroom/floor_light/bulb1', type 'attributeReport', cluster 'genOnOff', data '{"onOff":1}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2020-10-08 18:31:46: Received Zigbee message from 'livingroom/floor_light/bulb2', type 'attributeReport', cluster 'genOnOff', data '{"onOff":1}' from endpoint 1 with groupID 0
Zigbee2MQTT:info  2020-10-08 18:31:48: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/remote', payload '{"action":"on","battery":60,"linkquality":84,"update":{"state":"idle"}}'
Zigbee2MQTT:info  2020-10-08 18:31:48: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb1', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":57,"state":"ON"}'
Zigbee2MQTT:info  2020-10-08 18:31:48: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb2', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":65,"state":"ON"}'
Zigbee2MQTT:debug 2020-10-08 18:31:48: Received Zigbee message from 'livingroom/floor_light/bulb1', type 'attributeReport', cluster 'genLevelCtrl', data '{"currentLevel":254}' from endpoint 1 with groupID 0
Zigbee2MQTT:info  2020-10-08 18:31:50: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb1', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":57,"state":"ON"}'
Zigbee2MQTT:debug 2020-10-08 18:31:51: Received Zigbee message from 'livingroom/floor_light/bulb1', type 'attributeReport', cluster 'lightingColorCtrl', data '{"colorTemperature":555}' from endpoint 1 with groupID 0
Zigbee2MQTT:info  2020-10-08 18:31:53: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb1', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":60,"state":"ON"}'
```

```
Zigbee2MQTT:debug 2020-10-08 18:32:39: Received Zigbee message from 'livingroom/floor_light/remote', type 'commandOff', cluster 'genOnOff', data '{}' from endpoint 1 with groupID 10105
Zigbee2MQTT:debug 2020-10-08 18:32:39: Received Zigbee message from 'livingroom/floor_light/bulb1', type 'attributeReport', cluster 'genOnOff', data '{"onOff":0}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2020-10-08 18:32:40: Received Zigbee message from 'livingroom/floor_light/bulb2', type 'attributeReport', cluster 'genOnOff', data '{"onOff":0}' from endpoint 1 with groupID 0
Zigbee2MQTT:info  2020-10-08 18:32:41: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/remote', payload '{"action":"off","battery":60,"linkquality":57,"update":{"state":"idle"}}'
Zigbee2MQTT:info  2020-10-08 18:32:41: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb1', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":55,"state":"OFF"}'
Zigbee2MQTT:info  2020-10-08 18:32:42: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb2', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":42,"state":"OFF"}'
````

That was with a remote

```
Zigbee2MQTT:debug 2020-10-08 18:33:07: Received MQTT message on 'zigbee2mqtt/livingroom/floor_light/set/state' with data 'toggle'
Zigbee2MQTT:debug 2020-10-08 18:33:07: Publishing 'set' 'state' to 'livingroom/floor_light'
Zigbee2MQTT:debug 2020-10-08 18:33:07: Received Zigbee message from 'livingroom/floor_light/bulb1', type 'attributeReport', cluster 'genOnOff', data '{"onOff":1}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2020-10-08 18:33:07: Received Zigbee message from 'livingroom/floor_light/bulb2', type 'attributeReport', cluster 'genOnOff', data '{"onOff":1}' from endpoint 1 with groupID 0
Zigbee2MQTT:info  2020-10-08 18:33:07: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb1', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":42,"state":"OFF"}'
Zigbee2MQTT:info  2020-10-08 18:33:07: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"color_temp":555,"state":"OFF"}'
Zigbee2MQTT:info  2020-10-08 18:33:07: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb2', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":42,"state":"OFF"}'
Zigbee2MQTT:info  2020-10-08 18:33:09: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb1', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":42,"state":"OFF"}'
Zigbee2MQTT:info  2020-10-08 18:33:09: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb2', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":42,"state":"OFF"}'
```

```
Zigbee2MQTT:debug 2020-10-08 18:33:33: Received MQTT message on 'zigbee2mqtt/livingroom/floor_light/set/state' with data 'toggle'
Zigbee2MQTT:debug 2020-10-08 18:33:33: Publishing 'set' 'state' to 'livingroom/floor_light'
Zigbee2MQTT:debug 2020-10-08 18:33:33: Received Zigbee message from 'livingroom/floor_light/bulb1', type 'attributeReport', cluster 'genOnOff', data '{"onOff":0}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2020-10-08 18:33:33: Received Zigbee message from 'livingroom/floor_light/bulb2', type 'attributeReport', cluster 'genOnOff', data '{"onOff":0}' from endpoint 1 with groupID 0
Zigbee2MQTT:info  2020-10-08 18:33:33: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb1', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":42,"state":"ON"}'
Zigbee2MQTT:info  2020-10-08 18:33:33: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"color_temp":555,"state":"ON"}'
Zigbee2MQTT:info  2020-10-08 18:33:33: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb2', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":42,"state":"ON"}'
Zigbee2MQTT:info  2020-10-08 18:33:35: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb1', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":42,"state":"ON"}'
Zigbee2MQTT:info  2020-10-08 18:33:35: MQTT publish: topic 'zigbee2mqtt/livingroom/floor_light/bulb2', payload '{"brightness":254,"color":{"hue":360,"saturation":100,"x":0.6986,"y":0.2995},"linkquality":42,"state":"ON"}'
```

And using mqtt...

Neither seem to be correct? The one only coming in via reporting still don't update the group topic, while they should (there is nothing optimistic or oportunistic about it)

Using mqtt... now gets an update... but the state is wrong? it is ON when it should be off and off when it should be on!

**Comment by Koenkk** [comment]
> Neither seem to be correct? The one only coming in via reporting still don't update the group topic, while they should (there is nothing optimistic or oportunistic about it)

When optimistic is used for groups, an attribute report won't update the group state (it's documented here: https://www.zigbee2mqtt.io/information/groups.html#state-changes). This was implemented because of https://github.com/Koenkk/zigbee2mqtt/issues/764#issuecomment-531486974 (I agree `optimistic` might be a confusing name for this option).

**Comment by sjorge** [comment]
So there is no way to skip the update on set and only take reporting updates for groups?

**Comment by Koenkk** [comment]
Currently not indeed and the existing `optimistic` option makes this a bit harder to add. This should be fixed in Zigbee2MQTT 2.0.0 for sure (making optimistic for devices and groups the same).

**Comment by sjorge** [comment]
When this is revisited we should also fix debounce for groups, there is an issue for that too.

Do we have a list tracking these thing somewhere?

**Comment by Koenkk** [comment]
We don't, but should do and tag this with 'breaking change'. About the debounce for groups, I don't think this is breaking, so this can be implemented in Zigbee2MQTT 1.x?

**Comment by sjorge** [comment]
Yes, a tag would be nice... can we then also have the stalebot leave those alone? I assume those will be long lived.

https://github.com/Koenkk/zigbee2mqtt/issues/3461 is the debounce one, I had a brief looked but got super confused on how it works for devices.

**Comment by Koenkk** [comment]
Done

**Pull request merged**